.github/workflows/windows.yml

```diff
@@ -46,42 +46,163 @@ env:
+          gh api "$self_hosted_runner_scope" > $runners
+
+          # Generate a unique id that allows the workload job to find the runner
+          # we are reserving for it (via runner labels), and allows the timeout
+          # job to find the workload job run (via the job’s friendly name), even
+          # if there are multiple instances in the workflow call tree.
+          random32=$(LC_ALL=C < /dev/urandom tr -dC 0-9a-f 2> /dev/null | head -c 8)
+          unique_id=${{ github.repository }}/${{ github.run_id }}/$random32
+          echo "unique_id=$unique_id" | tee -a $GITHUB_OUTPUT
+
+          # Find a runner that is online, not busy, and not already reserved for
+          # any job (label prefix “reserved-for:”).
+          runner_id=$(mktemp)
+          if ! < $runners > $runner_id jq \
+            --arg self_hosted_image_label "$self_hosted_image_label" -e '
+              .runners
+              | map(select(.status == "online" and .busy == false))
+              | map(select([.labels[].name] | index($self_hosted_image_label) | not | not))
+              | map(select([.labels[].name] | map(startswith("reserved-for:")) | index(true) | not))
+              | first | .id'; then
+            fall_back_to_github_hosted
+          fi
+
+          # Reserve that runner by adding a label containing the unique id.
+          # Job concurrency ensures that runners never get assigned twice.
+          reserved_since=$(date +\%s)
+          gh api "$self_hosted_runner_scope/$(cat $runner_id)/labels" \
+            -f "labels[]=reserved-for:$unique_id" \
+            -f "labels[]=reserved-since:$reserved_since" \
+            --method POST --silent
+          echo "selected_runner_label=reserved-for:$unique_id" | tee -a $GITHUB_OUTPUT
+          echo 'is_self_hosted=true' | tee -a $GITHUB_OUTPUT
+
+  # In the unlikely event a self-hosted runner was selected and reserved but it
+  # goes down before the workload starts, cancel the workflow run.
+  runner-timeout:
+    needs:
+      - runner-select
+    if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) }}
+    name: Runner timeout
+    runs-on: ubuntu-latest
+    steps:
+      - name: Wait a bit
+        run: sleep 30
+
+      - name: Cancel if workload job is still queued
+        run: |
+          run_url=/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}
+          export GH_TOKEN=${{ secrets.GITHUB_TOKEN }}
+
+          if [ "$(gh api "$run_url/jobs" \
+            | jq -er --arg id '${{ needs.runner-select.outputs.unique-id }}' \
+              '.jobs[] | select(.name | contains("[" + $id + "]")) | .status'
+          )" = queued ]; then
+            echo "Timeout waiting for runner assignment! Cancelling workflow run"
+            gh api "$run_url/cancel" --method POST
+          fi
+
   build:
-    name: Windows Build
-    runs-on: windows-2022
+    needs:
+      - runner-select
+    name: Windows Build [${{ needs.runner-select.outputs.unique-id }}]
```

Comment:
NIT: This is empty when `fall_back_to_github_hosted`: https://github.com/sagudev/servo/actions/runs/10507052600/job/29108001368

Reply:
fixed in 7d0f50dd25c56b820efd6947e643a7570ef10235